consensus/src/runway/backup.rs

```diff
@@ -105,28 +117,41 @@ impl<R: Read, H: Hasher, D: Data, S: Signature> UnitLoader<R, H, D, S> {
 
 fn load_backup<H: Hasher, D: Data, S: Signature, R: Read>(
     unit_loader: UnitLoader<R, H, D, S>,
-) -> Result<(Vec<UncheckedSignedUnit<H, D, S>>, Round), LoaderError> {
-    let (mut rounds, units): (Vec<_>, Vec<_>) = unit_loader
+    index: NodeIndex,
+    session_id: SessionId,
+) -> Result<Vec<UncheckedSignedUnit<H, D, S>>, LoaderError> {
+    let (rounds, units): (Vec<_>, Vec<_>) = unit_loader
         .load()?
         .into_iter()
         .map(|u| (u.as_signable().coord().round(), u))
         .unzip();
-    rounds.sort_unstable();
-    if let Some(&max) = rounds.last() {
-        if max as usize + 1 < rounds.len() {
-            return Err(LoaderError::DuplicateRounds(max, rounds.len()));
+
+    for (round, expected_round) in rounds.iter().zip(0..) {
+        if *round != expected_round {
+            return Err(LoaderError::RoundMissmatch(expected_round, *round));
         }
-        if max as usize + 1 > rounds.len() {
-            return Err(LoaderError::MissingRounds(max, rounds.len()));
+    }
+
+    for u in units.iter() {
+        let su = u.as_signable();
+        let coord = su.coord();
+        if coord.creator() != index {
+            return Err(LoaderError::WrongCreator(
+                coord.round(),
+                index,
+                coord.creator(),
+            ));
         }
-        if !rounds.iter().cloned().eq((0..).take(rounds.len())) {
-            Err(LoaderError::MissingAndDuplicateRounds(rounds.len()))
-        } else {
-            Ok((units, max + 1))
+        if su.session_id() != session_id {
```

Comment:
This is already checked by the unit validator (later). It's a bit weird that you are double checking this one property here.

Reply:
I think it's reasonable for `load_backup` to return `Err` in case when the units come from a different session -- this is an error that (hopefully not, but) might happen. Otherwise the behavior would be pretty strange and hard to diagnose -- I mean the backup would be considered correct and only when "importing" the units we would realize they are nonsense and ignore them, right? So I would keep it.

Reply:
The problem is this can already happen with other unit properties, like e.g. a bitflipped signature, so I don't think just checking this one thing adds much value. Although I'm also not advocating strongly against this, I think all this (runway) should get refactored quite a bit to put unit validation in one place, not scattered here and there. Then this spot would contain validation anyway, and having this additional check now doesn't hurt much.

Reply:
I see this check similarly as the creator check. Both of them might happen by mistakenly giving a wrong file to validator. So I find this sanity check useful. I agree with the need of runway to be refactored, so maybe we can create a story for that (unless it is already created) and leave this check here for now? All this stuff should also be included in a task of refactor.

Reply:
I'm worried that a single "refactor runway" story will not be sufficiently detailed to actually produce useful results. I would rather go for something like "centralize unit validation", which is a much clearer goal. Other than that agreed, this should be a task.

Reply:
Then I will create a task for that (centralize unit validation)

Reply:
https://cardinal-cryptography.atlassian.net/browse/A0-886